In [192]:
import sklearn as sk
import numpy as np
from sklearn.datasets import fetch_20newsgroups
import pandas as pd
from sklearn.naive_bayes import MultinomialNB, BernoulliNB
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer


In [168]:
news = fetch_20newsgroups(subset='all')

In [169]:
#news

# 절차
#### 1. 데이터 형 파악

#### 2-1. 
###### TfidVector와 CounterVector의 성능비교
#### 2-2. 
###### 추청에서 MultinomialNB와 BernouNB의 성능비교

#### 3-1.  
###### 데이터에서 불필요한 데이터 처리
#### 3-2   
##### 2-1, 2-2 다시 적용

### News는 dict type으로 에는 다양한 데이터 정보들이 포함되고 있음
   ###### data: 실제 글이 적혀있음
   ##### target_names에는 분류명이 기재되어 있음
   ##### target은 해당 분류명의 분류 코드

#### 1) 데이터셋 속성 확인

In [170]:
print(news.keys())

dict_keys(['target_names', 'description', 'DESCR', 'data', 'target', 'filenames'])


In [171]:
print(news["filenames"][0])
print('==========================================\
=============')
print(news["target_names"][0])
print('==========================================\
=============')
print(news["description"][0])
print('==========================================\
=============')
#print(news["DESCR"][0])
print('==========================================\
=============')
print(news["data"][0])
print('==========================================\
=============')
print(news["target"][0])


C:\Users\cch88\scikit_learn_data\20news_home\20news-bydate-test\rec.sport.hockey\54367
alt.atheism
t
From: Mamatha Devineni Ratnam <mr47+@andrew.cmu.edu>
Subject: Pens fans reactions
Organization: Post Office, Carnegie Mellon, Pittsburgh, PA
Lines: 12
NNTP-Posting-Host: po4.andrew.cmu.edu



I am sure some bashers of Pens fans are pretty confused about the lack
of any kind of posts about the recent Pens massacre of the Devils. Actually,
I am  bit puzzled too and a bit relieved. However, I am going to put an end
to non-PIttsburghers' relief with a bit of praise for the Pens. Man, they
are killing those Devils worse than I thought. Jagr just showed you why
he is much better than his regular season stats. He is also a lot
fo fun to watch in the playoffs. Bowman should let JAgr have a lot of
fun in the next couple of games since the Pens are going to beat the pulp out of Jersey anyway. I was very disappointed not to see the Islanders lose the final
regular season game.          PENS RULE!!

In [172]:
print (news.description)

the 20 newsgroups by date dataset


In [173]:
print (type(news.data),type(news.target),type(news.target_names),type(news.DESCR),\
       type(news.filenames),type(news.description))

<class 'list'> <class 'numpy.ndarray'> <class 'list'> <class 'NoneType'> <class 'numpy.ndarray'> <class 'str'>


In [174]:
print(news.target_names)

['alt.atheism', 'comp.graphics', 'comp.os.ms-windows.misc', 'comp.sys.ibm.pc.hardware', 'comp.sys.mac.hardware', 'comp.windows.x', 'misc.forsale', 'rec.autos', 'rec.motorcycles', 'rec.sport.baseball', 'rec.sport.hockey', 'sci.crypt', 'sci.electronics', 'sci.med', 'sci.space', 'soc.religion.christian', 'talk.politics.guns', 'talk.politics.mideast', 'talk.politics.misc', 'talk.religion.misc']


In [179]:
print(news.data)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.


In [176]:
print(news.target)

[10  3 17 ...,  3  1  7]


In [177]:
(news.filenames)

array([ 'C:\\Users\\cch88\\scikit_learn_data\\20news_home\\20news-bydate-test\\rec.sport.hockey\\54367',
       'C:\\Users\\cch88\\scikit_learn_data\\20news_home\\20news-bydate-train\\comp.sys.ibm.pc.hardware\\60215',
       'C:\\Users\\cch88\\scikit_learn_data\\20news_home\\20news-bydate-train\\talk.politics.mideast\\76120',
       ...,
       'C:\\Users\\cch88\\scikit_learn_data\\20news_home\\20news-bydate-train\\comp.sys.ibm.pc.hardware\\60695',
       'C:\\Users\\cch88\\scikit_learn_data\\20news_home\\20news-bydate-train\\comp.graphics\\38319',
       'C:\\Users\\cch88\\scikit_learn_data\\20news_home\\20news-bydate-test\\rec.autos\\103195'],
      dtype='<U95')

In [180]:
print(news.description)

the 20 newsgroups by date dataset


In [181]:
print(len(news.target),len(news.description),len(news.data),len(news.target_names),len(news.filenames))

18846 33 18846 20 18846


In [182]:
news.description[0:]

'the 20 newsgroups by date dataset'

###### data - list형 len=18846 
###### target - ndarry형 len=18846 
###### target_names - list형 len=20
###### DESCR - none형  
###### Filenames - ndarry형 len=18846  
###### description - str형 len=30(단순 문장 - 'the 20 newsgroups by date dataset')

In [183]:
print (news.data[0],news.target[0],news.target_names[news.target[0]],news.filenames[0])

From: Mamatha Devineni Ratnam <mr47+@andrew.cmu.edu>
Subject: Pens fans reactions
Organization: Post Office, Carnegie Mellon, Pittsburgh, PA
Lines: 12
NNTP-Posting-Host: po4.andrew.cmu.edu



I am sure some bashers of Pens fans are pretty confused about the lack
of any kind of posts about the recent Pens massacre of the Devils. Actually,
I am  bit puzzled too and a bit relieved. However, I am going to put an end
to non-PIttsburghers' relief with a bit of praise for the Pens. Man, they
are killing those Devils worse than I thought. Jagr just showed you why
he is much better than his regular season stats. He is also a lot
fo fun to watch in the playoffs. Bowman should let JAgr have a lot of
fun in the next couple of games since the Pens are going to beat the pulp out of Jersey anyway. I was very disappointed not to see the Islanders lose the final
regular season game.          PENS RULE!!!

 10 rec.sport.hockey C:\Users\cch88\scikit_learn_data\20news_home\20news-bydate-test\rec.sport.hoc

In [184]:
news_df = pd.DataFrame({'News' : news.data,
                       'Target' : news.target})

In [185]:
news_df.head()

News  Target
0  From: Mamatha Devineni Ratnam <mr47+@andrew.cm...      10
1  From: mblawson@midway.ecn.uoknor.edu (Matthew ...       3
2  From: hilmi-er@dsv.su.se (Hilmi Eren)\nSubject...      17
3  From: guyd@austin.ibm.com (Guy Dawson)\nSubjec...       3
4  From: Alexander Samuel McDiarmid <am2o+@andrew...       4

#### 데이터  전처리
##### 테스트,트레이닝 구분

In [186]:
SPLIT_Ratio=0.75
split_size = int(len(news.data)* SPLIT_Ratio)
X_train = news_df['News'][:split_size]
X_test = news_df['News'][split_size:]
Y_train = news_df['Target'][:split_size]
Y_test = news_df['Target'][split_size:]

### 파이프라인 형성

In [187]:
multiNB_1 = Pipeline([\
                      ('vect',CountVectorizer()),
                     ('clf',MultinomialNB()),
                     ])#파이프 라인은 투플 형식으로 중간단계는 사이킷런 변환기, 마지막은 에스티메이터이다.

multiNB_2 = Pipeline([\
                      ('vect',TfidfVectorizer()),
                     ('clf',MultinomialNB()),
                     ])

BernouNB_1 = Pipeline([\
                      ('vect',CountVectorizer()),
                     ('clf',BernoulliNB()),
                     ])

BernouNB_2 = Pipeline([\
                      ('vect',TfidfVectorizer()),
                     ('clf',BernoulliNB()),
                     ])


In [188]:
from sklearn.cross_validation import cross_val_score,KFold

In [193]:
def evaluate_cross_validatio(clf,X,Y,K):
    #k=5 교차검증
    cv = KFold(len(Y), n_folds=K, shuffle=True, random_state=0)
    
    scores = cross_val_score(clf,X,Y,cv=cv)
    print (scores)
    print('Mean score: %.3f +/- %.3f' % (np.mean(scores),np.std(scores)))

In [194]:
clfs = [multiNB_1,multiNB_2,BernouNB_1,BernouNB_2]

In [195]:
for clf in clfs:
    evaluate_cross_validatio(clf,X_train,Y_train,5)
    
    

[ 0.83445348  0.84471171  0.83657588  0.84258932  0.83581033]
Mean score: 0.839 +/- 0.004
[ 0.82950124  0.83091617  0.82808631  0.85037142  0.82236377]
Mean score: 0.832 +/- 0.010
[ 0.6681995   0.67739653  0.66501592  0.66536965  0.64401982]
Mean score: 0.664 +/- 0.011
[ 0.6681995   0.67739653  0.66501592  0.66536965  0.64401982]
Mean score: 0.664 +/- 0.011


###### CounterVectorize와 TfidVectorizer는 성능이 비슷하다 과제 조건에서 되도록 TfidVectorizer를 권장하여서 이후 과정에는 TfidVectorizer만 사용
   ######     다만 에스티메이터에서  베르누이분포와  Multinomial를  비교하면  Mutinomial의 성능이 더 우수하다.
  
   

##### 정규화를 통한 데이터 정제 (전영호 형님꺼 인용하였습니다.)


In [45]:
import re

In [46]:
def data_cleansing(text):
    delete_email = re.sub(r'\b[\w\+]+@[\w]+.[\w]+.[\w]+.[\w]+\b', ' ', text)
    delete_number = re.sub(r'\b|\d+|\b', ' ',delete_email)
    delete_non_word = re.sub(r'\b[\W]+\b', ' ', delete_number)
    cleaning_result = ' '.join(delete_non_word.split())
    return cleaning_result

In [47]:
news_df['News'] = news_df['News'].apply(data_cleansing)

In [142]:
 evaluate_cross_validatio(multiNB_2,news_df['News'],news_df['Target'],5)

[ 0.8397878   0.85699124  0.84266384  0.85486867  0.84611303]
Mean score: 0.848 +/- 0.007


#####  정규화를 통해 0.018정도 평균점수상승 한것으로 볼 수 있다.

In [143]:
SPLIT_Ratio=0.75
split_size = int(len(news.data)* SPLIT_Ratio)
X_train = news_df['News'][:split_size]
X_test = news_df['News'][split_size:]
Y_train = news_df['Target'][:split_size]
Y_test = news_df['Target'][split_size:]


###### 포터 스테머 알고리즘 - 토큰화의 일종으로 단어의 파생어들을 원형으로 정제시켜준다.


In [86]:
from nltk.stem.porter import PorterStemmer
porter = PorterStemmer()

In [87]:

def tokenizer_porter(text):
    return [porter.stem(word) for word in text.split()]


###### 불용어 처리 - stop words ====> ex) is, and, has 등 처리 안 불러와진다 ㅠㅠㅠ 대신 TfidVectiorizer 자체것으로 씀

In [35]:
import nltk

In [41]:
nltk.download('stopwords')

[nltk_data] Error loading stopwords: HTTP Error 405: Not allowed.


False

In [42]:
from nltk.corpus import stopwords

AttributeError: 'list' object has no attribute 'apply'

### 정규화, 토큰화를 거친 파이프라인

In [156]:
new_pipe = Pipeline([('vect',TfidfVectorizer(stop_words='english',)),
                     ('clf',MultinomialNB(0.01)),])
new_pipe2 = Pipeline([('vect',TfidfVectorizer(stop_words='english',)),
                      ('clf',MultinomialNB()),])
#new_pipe2 = Pipeline([('vect',CountVectorizer(stop_words='english', token_pattern=tokenizer_porter,)),
                      #('clf',MultinomialNB()),])   

### alpha는 평활화로 연관된 단어가 나타날 확률을 0이 아닌 값으로 설정한다. 즉, 가짜 빈도수
###### 예로 들면 맥주라는 단어는 스포츠 분류단어와 관련이 없다. 이럴 때 맥주라는 단어는 스포츠가 아닌 단어가 된다. 
###### 즉,스포츠의 하키 기사일 때 맥주와 관련 있을 확률이 0이 된다. "하키 경기장에서 관중이 맥주 투척" 문장이 나온다면 
###### 맥주 단어가 들어가 있기에 항상 스포츠와 관련 없다고 간주된다. 따라서  가짜확률을 집어 넣어 관련없어보이는 
###### 맥주라는 단어가 나올 때 스포츠의 하키와 관련있을 확률을 대체한다.


In [157]:
clfs=[new_pipe,new_pipe2]

In [158]:
for clf in clfs:
    evaluate_cross_validatio(clf,X_train,Y_train,5)

[ 0.90802971  0.90696852  0.90909091  0.91651928  0.90304317]
Mean score: 0.909 +/- 0.004
[ 0.85921472  0.86416696  0.8712416   0.87407145  0.86553432]
Mean score: 0.867 +/- 0.005


* 실제로 0.042정도 성능이 증가하였다.

##### 포터스테머 알고리즘 tokenizer를 사용

In [200]:
new_pipe3 = Pipeline([('vect',TfidfVectorizer(stop_words='english',tokenizer=tokenizer_porter)),
                     ('clf',MultinomialNB(0.01)),])

In [201]:
evaluate_cross_validatio(new_pipe3,X_train,Y_train,5)


KeyboardInterrupt: 

* 사실상 성능의 차이가 없다. 시간이 상당히 오래걸린다. 파생어들을 어근으로 처리하는 시간이 걸린다.

#### 모델의 성능 평가

In [223]:
from sklearn import metrics

In [224]:
def train_and_evaluate(clf,X_train,X_test,Y_train,Y_test):
    clf.fit(X_train,Y_train)
    print ("Accuracy training set:")
    print(clf.score(X_train,Y_train))
    print ("Accuracy testing set:")
    print(clf.score(X_test,Y_test))
    Y_pred = clf.predict(X_test)
    
    
    print ("Classification Report")
    print (metrics.classification_report(Y_test,Y_pred))
    print ("Confusion Matrix")
    print (metrics.confusion_matrix(Y_test,Y_pred))

In [225]:
train_and_evaluate(new_pipe3,X_train,X_test,Y_train,Y_test)

Accuracy training set:
0.998089712749
Accuracy testing set:
0.914049235993
Classification Report
             precision    recall  f1-score   support

          0       0.93      0.89      0.91       216
          1       0.82      0.81      0.81       246
          2       0.89      0.85      0.87       274
          3       0.78      0.86      0.82       235
          4       0.89      0.87      0.88       231
          5       0.88      0.92      0.90       225
          6       0.86      0.83      0.85       248
          7       0.95      0.93      0.94       275
          8       0.95      0.97      0.96       226
          9       0.98      0.95      0.97       250
         10       0.98      1.00      0.99       257
         11       0.96      0.97      0.97       261
         12       0.90      0.91      0.91       216
         13       0.95      0.94      0.94       257
         14       0.95      0.93      0.94       246
         15       0.89      0.96      0.93       234
 